### ISSUE TITLE
Subscriptable on Option[MyClass] not ruled out by and/or Eliminations

### ISSUE THREAD

**Issue by KoldenAxelson** [microsoft/pyright#9548]
**Describe the bug**
```
Pyright: Object of type "None" is not subscriptable
self[direction][direction] 
^^^^^^^^^^^^^^^
```
Since self[direction] can not be None by the time of this check, it's should be a subscriptable Cell.
The code runs just fine, Pyright is just incorrectly marking the line.

**Code**
```python
from enum import Enum
from dataclasses import dataclass

class Direction(Enum):
    UP    = 'up'
    RIGHT = 'right'
    DOWN  = 'down'
    LEFT  = 'left'

@dataclass
class Cell():
    # Properties
    value: str
    up   : Optional['Cell']
    right: Optional['Cell']
    down : Optional['Cell']
    left : Optional['Cell']
    # Dunder Methods
    def __init__(self, value: str) -> None:
        self.value = value
        self.up, self.right, self.down, self.left = None, None, None, None
    def __getitem__(self, direction: Direction) -> Optional['Cell']:
        match direction:
            case Direction.UP   : return self.up
            case Direction.RIGHT: return self.right
            case Direction.DOWN : return self.down
            case Direction.LEFT : return self.left
            case _              : return None
    def __str__(self) -> str:
        return self.value
    # Public Methods
    def is_match(self) -> bool:
        return False
    # Private Methods
    def _is_X(self) -> bool:
        return self.value == 'X'
    def _match_direction(self,direction:Direction) -> bool:
        if self[direction] is None or \
            self[direction][direction] is None or \
            self[direction][direction][direction] is None: return False
        if f'{self[direction]}' != 'M' or \
            f'{self[direction][direction]}' != 'A' or \
            f'{self[direction][direction][direction]}' != 'S': return False
        return True

if __name__ == "__main__":
    cell_X: Cell = Cell('X')
    cell_M: Cell = Cell('M'); cell_X.down = cell_M
    cell_A: Cell = Cell('A'); cell_M.down = cell_A
    cell_S: Cell = Cell('S'); cell_A.down = cell_S

    print(cell_X._match_direction(Direction.DOWN)) # This returns True
```

**VS Code extension or command-line**
I'm currently using Pyright in the current version of Zed (12/04/2024)

**Comment by erictraut** [comment]
Pyright is working as intended here, so I don't consider this a bug.

Pyright doesn't narrow indexed expressions where the subscript is a variable. This is also true of other static type checkers like mypy.

For more details about type narrowing support in pyright, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards).

If you want to use this pattern and avoid type checking errors, you can use an intermediate variable assigned with a walrus operator.

```python
    def _match_direction(self, direction: Direction) -> bool:
        if (x := self[direction]) is None or (y := x[direction]) is None or y[direction] is None:
            return False
        if f"{x := self[direction]}" != "M" or f"{y := x[direction]}" != "A" or f"{y[direction]}" != "S":
            return False
        return True
```
